overlord/snapshotstate/backend/backend.go

```diff
@@ -429,73 +421,66 @@ func addDirToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer,
 		// order to do this, we return a "-" as a sentinel.
 		return "-"
 	}
+
+	var expExcludePaths []string
 	for _, excludePath := range excludePaths {
 		expandedPath := os.Expand(excludePath, expandSnapDataDirs)
 		// "-" is the sentinel returned by expandSnapDataDirs() if the
 		// exclusion path is not relevant for the type of data being considered
 		if expandedPath[0] == '-' {
 			continue
 		}
-		tarArgs = append(tarArgs, fmt.Sprintf("--exclude=%s", expandedPath))
+		expExcludePaths = append(expExcludePaths, expandedPath)
 	}
 
-	noRev, noCommon := true, true
+	return addToZip(ctx, snapshot, w, username, entry, paths, expExcludePaths)
+}
 
-	exists, isDir, err = osutil.DirExists(dir)
+// addToZip adds 'paths' to the snapshot. tar will change into the paths' parent
+// directory before creating the archive so that parent dirs are not added.
+func addToZip(ctx context.Context, snapshot *client.Snapshot, w *zip.Writer, username, entry string, paths []string, excludePaths []string) error {
+	archiveWriter, err := w.CreateHeader(&zip.FileHeader{Name: entry})
 	if err != nil {
 		return err
 	}
-	switch {
-	case exists && isDir:
-		tarArgs = append(tarArgs, revdir)
-		noRev = false
-	case exists && !isDir:
-		logger.Noticef("Not saving %q in snapshot #%d of %q as it is not a directory.", dir, snapshot.SetID, snapshot.Snap)
-	case !exists:
-		logger.Debugf("Not saving %q in snapshot #%d of %q as it is does not exist.", dir, snapshot.SetID, snapshot.Snap)
-	}
 
-	common := filepath.Join(parent, "common")
-	exists, isDir, err = osutil.DirExists(common)
-	if err != nil {
-		return err
-	}
-	switch {
-	case exists && isDir:
-		tarArgs = append(tarArgs, "common")
-		noCommon = false
-	case exists && !isDir:
-		logger.Noticef("Not saving %q in snapshot #%d of %q as it is not a directory.", common, snapshot.SetID, snapshot.Snap)
-	case !exists:
-		logger.Debugf("Not saving %q in snapshot #%d of %q as it is does not exist.", common, snapshot.SetID, snapshot.Snap)
+	tarArgs := []string{
+		"--create",
+		"--sparse", "--gzip",
+		"--format", "gnu",
+		"--anchored",
+		"--no-wildcards-match-slash",
 	}
 
-	if noCommon && noRev {
-		return nil
+	for _, path := range excludePaths {
+		tarArgs = append(tarArgs, fmt.Sprintf("--exclude=%s", path))
 	}
 
-	archiveWriter, err := w.CreateHeader(&zip.FileHeader{Name: entry})
-	if err != nil {
-		return err
+	// use --directory so that the file is added without its parent dirs
+	for _, path := range paths {
+		parent, file := filepath.Split(path)
+		tarArgs = append(tarArgs, "--directory", parent, file)
 	}
```

Comment:
It's not immediately obvious why this is correct. We know that we'll have at most two paths, and both have the same parent, but we should probably write it as a comment both here and in the function docstring (to make clear that it's a precondition).
And here you could add a `break`, since we want the loop to be run just once.

Reply:
I'm not sure I agree/understand. In this function, we don't know we'll have at most two paths. That's an implementation detail specific to the structure of the snap user dir which this refactor tries to move out of this function and into the one above (in fact, that won't be true when we add ~/Snap). Also, could you explain the addition of a `break` here? AFAICT, running this once would mean we wouldn't add one of the subdirectories (the revisioned or the common one).

Reply:
Ah! I missed the `file` parameter at the end, so I thought that this code was just adding the `--directory parent` options one after the other, and in that case only the last one matters. But indeed, you are doing it right. :-)

Then maybe a very minor nitpick would be to rename the local variable `file` to `dir` or `dirname`.

Reply:
Done. Sorry, I marked it with a +1 but then forgot to do it :sweat_smile: